### PR TITLE
Giphy integration

### PR DESCRIPTION
**Pull request by panoschal**
First of all, this needs a lot of input from product and design.

Fixes vector-im/element-meta#321 !

# Flow

I created a new button in MessageComposer, next to the emoji picker and the sticker picker.

I consume the Giphy API for search results, which should be better than the SDK due to privacy concerns, and possible changes in the future.

I display the GIF search results in an one-column list.

I added an option in labs for showing the GIF picker.

image

I send the GIF as an attachment. Another option is to just send the url, but previews generated by the server have some issues currently and it doesn't work: https://github.com/matrix-org/synapse/issues/1278

# TODO
- [x] add svg icon
- [x] test in electron
- [x] test in white/dark themes
- [x] pick correct, small enough filesize to send
- [ ] convert list to two-column?
- [x] disable confirmation dialog
- [x] add option in labs
- [ ] add "powered by Giphy"
- [ ] show trending gifs when search box is empty?
- [ ] (for core devs) create api key from https://developers.giphy.com/dashboard/

PR for documentation in labs.md: https://github.com/vector-im/element-web/pull/16808

Related issue: https://github.com/vector-im/element-web/issues/2693

Signed-off-by: Panagiotis Chalimourdas <panagiotis4531@gmail.com>

---
Here's what your changelog entry will look like:

## ✨ Features
 * Giphy integration ([\#5814](https://github.com/matrix-org/matrix-react-sdk/pull/5814)). Fixes vector-im/element-meta#321. Contributed by @panoschal.

Preview: https://613dd49f7161e91e7bfbfbbf--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.

### CODE REVIEW

**Comment by SimonBrandner** [comment]
First off, this is awesome :D

Fixes [#415](https://github.com/vector-im/element-web/issues/415)

> add option in labs?

I am pretty sure that will be required

**Comment by ShadowJonathan** [comment]
Why is this in `matrix-react-sdk`? I see this purely as an addition to matrix in the regards of Element, so this should (in my opinion) only be implemented from element-web.

I think there's a bit of a separation-of-concerns issue here.

**Comment by jryans** [comment]
> Why is this in `matrix-react-sdk`? I see this purely as an addition to matrix in the regards of Element, so this should (in my opinion) only be implemented from element-web.
> 
> I think there's a bit of a separation-of-concerns issue here.

Hmm, it's a bit unclear perhaps. I think we should first review the product fit here, and for there we can work out the proper home for the code.

**Comment by kojid0** [comment]
Why not use the sidebar? There is enough space to display many things more clearly than in a small window. This also applies to emoji and stickers... GIFs could be the first to support sidebar accesibility, with Emoji and Sticker implemented later.

111 (Custom)

**Comment by SimonBrandner** [comment]
> Why not use the sidebar? There is enough space to display many things more clearly than in a small window. This also applies to emoji and stickers... GIFs could be the first to support sidebar accesibility, with Emoji and Sticker implemented later.

This wouldn't match the current design... So if this change were to be implemented, it should in bulk with emojis and stickers. This is ultimately up to the design team to decide. I remember there being an issue about this, but I can't find it

**Comment by panoschal** [comment]
There is this issue https://github.com/vector-im/element-web/issues/16180 and this one https://github.com/vector-im/element-web/issues/2693#issuecomment-762741358

We can also have the popup panel bigger, with the size of the emoji picker or bigger, and have two columns of gifs.

**Comment by tsukasagenesis** [comment]
nice feature ! can't wait to get it integrated !

File: src/components/views/gifpicker/GifPicker.tsx [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+import React from "react";
+import { replaceableComponent } from "../../../utils/replaceableComponent";
+import Search from "../emojipicker/Search";
+import ScrollPanel from "../../structures/ScrollPanel";
+import Spinner from "../elements/Spinner";
+import GifThumbnail from "./GifThumbnail";
+import { DebouncedFunc, throttle, uniqBy } from "lodash";
+import { Gif } from "./Gif";
+
+const API_KEY = "hhMBhTnD7k8BgyoZCM9UNM1vRsPcgzaC";
```
Comment (annie-elequin) [inline src/components/views/gifpicker/GifPicker.tsx:11]: This probably shouldn't be checked into git. 
Unrelated though, this work is amazing :)

Comment (panoschal) [reply src/components/views/gifpicker/GifPicker.tsx:11]: Thanks!

This is a public key so it is not a security issue, as the request will be done client-side and anyone can inspect it.
https://github.com/Giphy/giphy-js/issues/120#issuecomment-680988917

If this is approved by the product team a production key is needed in order to increase the limits by Giphy.

**Comment by t3chguy** [comment]
Removing review as PR is a draft

**Comment by panoschal** [comment]
> Removing review as PR is a draft

@t3chguy this needs feedback by the product team in order to continue development, so should I mark it as ready for review?

**Comment by t3chguy** [comment]
Yes, draft means still work in progress, not ready for review.

image

**Comment by panoschal** [comment]
I will wait for feedback, then, before continuing with what I think could be done because the feature may not be desired in this place.

**Comment by erAck** [comment]
If this hits product then please

- make it a preference option
- have it disabled per default
- have it disabled per default for recipients (i.e. do _not_ display the gif unless enabled)
- finally have a way to fold away images in rooms

Reason is that animated gifs are utterly distracting, easily get on the nerves and unnecessarily consume CPU and power .
Also, images in general take too much space in the timeline, which the last bullet point would take care of.

Additionally, Giphy is privacy invading as the client's IP and search ends up with Facebook that as we all know will collect everything it can get its hands on.

I actually wish this will never get implemented.

**Comment by seanfarley** [comment]
> If this hits product then please
> 
> * make it a preference option
> * have it disabled per default
> * have it disabled per default for recipients (i.e. do _not_ display the gif unless enabled)
> * finally have a way to fold away images in rooms
> 
> Reason is that animated gifs are utterly distracting, easily get on the nerves and unnecessarily consume CPU and power .
> Also, images in general take too much space in the timeline, which the last bullet point would take care of.
> 
> Additionally, Giphy is privacy invading as the client's IP and search ends up with Facebook that as we all know will collect everything it can get its hands on.
> 
> I actually wish this will never get implemented.

Tell me you're a toxic dev without *telling me* you're a toxic dev.

**Comment by DoM1niC** [comment]
@panoschal can you fix the conflicts PLZ :) I wanne use this in my Fork

**Comment by panoschal** [comment]
I fixed the conflicts @DoM1niC

**Comment by DoM1niC** [comment]
Nice and THX I will try it out soon 😊😊

**Review by turt2live — COMMENTED** [review]
Thanks for the PR! This is quite a large chunk of functionality that, as you noted, needs design + product review. I'm mostly leaving this comment to clear the automatic code review request that github will insist on while we wait for review (given code review on something this size before design looks at it can be detrimental to progress). 

Something I did notice is that the new files are missing copyright/license headers - these should be added to ensure everything is all squared away :)

**Comment by DoM1niC** [comment]
> Great, I will try to implement these improvements soon.

needs threads support the Composer isn't upload the gif there... :(

**Comment by nnuel** [comment]
Just tested, I like the uploading feature, and would also second the two-column lay-out.

**Comment by williamkray** [comment]
Rather than including the giphy api key in the code, this should be something users can fill in, or a config value pulled from the element-web config.json. 

If the value is not supplied, the feature should be hidden/nonfunctional.

**Comment by arkitoure** [comment]
What's the status of this integration? Just throwing this out into the ether...how about selectable support for other gif platforms?

- tenor.com
- gifer.com
- gfycat.com
- duckduckgo gif search

**Comment by floviolleau** [comment]
Hi,

Any news on this awesome integration ready to fly?

Thanks!

**Comment by iceHtwoO** [comment]
Hi,
due to this PR not being updated, I have started developing a solution myself.
It should be ready in a few days.
Screenshot 2022-05-27 141323

**Comment by turt2live** [comment]
Thank you for working on this, however per https://github.com/vector-im/element-meta/issues/321#issuecomment-1140305492 I don't think we can safely keep this sort of change open. The feature requires design and product review and is generally quite large in size - we might try to take the work on ourselves if/when product/design get a chance to take a look.

**Comment by khazaddum** [comment]
I conviced my family to switch to a Matrix homeserver and Element, and their first question was "how do I send GIFs?"

**Comment by CRCinAU** [comment]
Did this ever get anywhere?

Or has it become a wasteland where PRs go to die?

**Comment by iceHtwoO** [comment]
> Did this ever get anywhere?
> 
> 
> 
> Or has it become a wasteland where PRs go to die?

This is sadly going no where, you can find more information in the following [issue](https://github.com/vector-im/element-meta/issues/321). There have been multiple attempts to integrate the feature.

**Comment by WesLee91** [comment]
Anyone know of any Matrix client that has giphy intergration?

**Pull request closed without merging**